doubly_linked_list/doubly_linked_list.py

```diff
@@ -1,68 +1,163 @@
-  def insert_before(self, value):
-    current_prev = self.prev
-    self.prev = ListNode(value, current_prev, self)
-    if current_prev:
-      current_prev.next = self.prev
 
-  """Rearranges this ListNode's previous and next pointers
+    def insert_before(self, value):
+        current_prev = self.prev
+        self.prev = ListNode(value, current_prev, self)
+        if current_prev:
+            current_prev.next = self.prev
+
+    """Rearranges this ListNode's previous and next pointers
   accordingly, effectively deleting this ListNode."""
-  def delete(self):
-    if self.prev:
-      self.prev.next = self.next
-    if self.next:
-      self.next.prev = self.prev
+
+    def delete(self):
+        if self.prev:
+            self.prev.next = self.next
+        if self.next:
+            self.next.prev = self.prev
+
 
 """Our doubly-linked list class. It holds references to
 the list's head and tail nodes."""
+
+
 class DoublyLinkedList:
-  def __init__(self, node=None):
-    self.head = node
-    self.tail = node
-    self.length = 1 if node is not None else 0
+    def __init__(self, node=None):
+        self.head = node
+        self.tail = node
+        self.length = 1 if node is not None else 0
+
+    def __len__(self):
+        return self.length
+
+    def add_to_head(self, value):
+        current_head = self.head
+        new_head = ListNode(value)
+
+        if not current_head:
+            self.head = new_head
+        else:
+            current_head.insert_before(value)
+            self.head = current_head.prev
+            self.length += 1
+
+    def remove_from_head(self):
+
+        if not self.head and not self.tail:
+            return None
+
+        if self.head.next is None:
+            removed_head = self.head.value
+            self.head = None
+            self.tail = None
+            self.length -= 1
+            return removed_head
+
+        else:
+            removed_head = self.head.value
+            self.head = self.head.next
+            self.head.prev = None
+            self.length -= 1
+
+            return removed_head
 
-  def __len__(self):
-    return self.length
+    def add_to_tail(self, value):
+        if self.tail is None:
+            new_node = ListNode(value)
+            self.head = new_node
+            self.tail = new_node
+        else:
+            self.tail.insert_after(value)
+            self.tail = self.tail.next
 
-  def add_to_head(self, value):
-    pass
+        self.length += 1
 
-  def remove_from_head(self):
-    pass
+    def remove_from_tail(self):
+        if not self.tail:
+            return None
+        removed_tail = self.tail.value
+        if self.tail == self.head:
+            self.head = None
+            self.tail = None
+        else:
+            self.tail.delete()
+            self.tail = self.tail.prev
+        self.length -= 1
+        return removed_tail
 
-  def add_to_tail(self, value):
-    pass
+    def move_to_front(self, node):
```

Comment:
move to front is a bit more complicated than moving to the end, aye? this is true in general for programming and and data structs. remember this concept; it will help tremendously later on :D